File: 2020/day15.py
INPUT = [14, 3, 1, 0, 9, 5]


def print_ans(input, n):
    prev_last_usage = {}
    last_usage = {}
    for i, curr_num in enumerate(input):
        last_usage[curr_num] = i
    for i in range(len(input), n):
        if curr_num not in prev_last_usage:
            curr_num = 0
        else:
            curr_num = last_usage[curr_num] - prev_last_usage[curr_num]
        if curr_num in last_usage:
            prev_last_usage[curr_num] = last_usage[curr_num]
        last_usage[curr_num] = i
    print(curr_num)


print_ans(INPUT, 30000000)

File: 2020/day15/day15.rb
INPUT = [14, 3, 1, 0, 9, 5]


def print_ans(input, num_rounds)
  last_usage = Array.new(num_rounds) { -1 }
  (input.length - 1).times do |i|
    last_usage[input[i]] = i
  end

  prev_num = input.last
  curr_num = nil
  (input.length..(num_rounds - 1)).each do |i|
    if last_usage[prev_num] >= 0
      curr_num = i - 1 - last_usage[prev_num]
    else
      curr_num = 0
    end
    last_usage[prev_num] = i - 1
    prev_num = curr_num
  end
  puts curr_num
end

print_ans(INPUT, 30000000)
